### PR TITLE
Add a crate for the nice control API

### PR DESCRIPTION
**Pull request by badboy**
For now this only supports a minimal interface on top of Glean

```rust
let cfg = Configuration {
    data_path,
    application_id: "org.mozilla.glean_core.example".into(),
    upload_enabled: true,
    max_events: None,
};
glean::initialize(cfg)?;
glean::register_ping_type(&PrototypePing);

if glean::send_ping_by_name("prototype") {
    log::info!("Successfully collected a prototype ping");
} else {
    log::info!("Prototype ping failed");
}
```

### CODE REVIEW

File: glean-core/preview/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+[package]
+name = "glean-preview"
+version = "0.0.1"
+authors = ["Jan-Erik Rediger <jrediger@mozilla.com>", "The Glean Team <glean-team@mozilla.com>"]
+description = "Nice Glean Rust API"
```
Comment (Dexterp37) [inline glean-core/preview/Cargo.toml]:
```suggestion
description = "Nice Glean SDK Rust API"
```

File: glean-core/preview/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+repository = "https://github.com/mozilla/glean"
+readme = "README.md"
+license = "MPL-2.0"
+edition = "2018"
+keywords = ["telemetry"]
```
Comment (Dexterp37) [inline glean-core/preview/Cargo.toml]:
```suggestion
keywords = ["telemetry", "glean"]
```

?

File: glean-core/preview/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  "Cargo.toml",
+]
+
+[badges]
+circle-ci = { repository = "mozilla/glean", branch = "master" }
+maintenance = { status = "actively-developed" }
+
+[dependencies.glean-ffi]
+path = "../ffi"
+version = "21.2.0"
```
Comment (Dexterp37) [inline glean-core/preview/Cargo.toml]:
```suggestion
version = "21.3.0"
```

File: glean-core/preview/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+[dependencies.glean-core]
+path = ".."
+version = "21.2.0"
```
Comment (Dexterp37) [inline glean-core/preview/Cargo.toml]:
```suggestion
version = "21.3.0"
```

File: glean-core/preview/examples/prototype.rs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+use std::env;
```
Comment (Dexterp37) [inline glean-core/preview/examples/prototype.rs]: Please add a license header here.

File: glean-core/preview/Cargo.toml (hunk 2 above)
Comment (badboy) [reply glean-core/preview/Cargo.toml]: Glean isn't a keyword by itself yet (https://crates.io/keywords/glean), but I guess I should start adding it to the other crates.

File: glean-core/preview/Cargo.toml (hunk 3 above)
Comment (badboy) [reply glean-core/preview/Cargo.toml]: Fun fact: it actually pulls in 21.3.0 ;)

File: glean-core/preview/examples/prototype.rs (hunk 5 above)
Comment (badboy) [reply glean-core/preview/examples/prototype.rs]: Woops, that one got lost in a rebase, adding it back.

File: glean-core/preview/Cargo.toml (hunk 2 above)
Comment (Dexterp37) [reply glean-core/preview/Cargo.toml]: Ah! Will it be shown once more published packages start using the keyword?

Comment (badboy) [reply glean-core/preview/Cargo.toml]: It should, yes

**Comment by badboy** [comment]
> Does [this](https://mozilla.github.io/glean/book/dev/cut-a-new-release.html) need to change or do you want to update [the script](https://github.com/mozilla/glean/blob/4915dbc6be78fb8669ea42bd3c60c16591bd4005/bin/prepare-release.sh#L78) ?

IMO this should not yet be part of the regular release cycle. It's preview after all and we might just scrape it by January

File: glean-core/preview/Cargo.toml (hunk 2 above)
Comment (Dexterp37) [reply glean-core/preview/Cargo.toml]: Fine by me, then :)

File: glean-core/preview/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+pub fn set_upload_enabled(flag: bool) -> bool {
+    with_glean(|glean_handle| {
+        glean_ffi::glean_set_upload_enabled(glean_handle, flag as u8);
+        glean_ffi::glean_is_upload_enabled(glean_handle) != 0
+    })
+}
+
+/// Determine whether upload is enabled.
+///
+/// See `glean_core::Glean.set_upload_enabled`.
```
Comment (chutten) [inline glean-core/preview/src/lib.rs]: copy-pasta?

File: glean-core/preview/src/metrics/ping.rs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+use ffi_support::FfiStr;
+use std::ffi::CString;
+
+/// Stores information about a ping.
+///
+/// This is required so that given metric data queued on disk we can send
+/// pings with the correct settings, e.g. whether it has a client_id.
+#[derive(Clone, Debug)]
+//pub struct PingType(pub(crate) glean_core::metrics::PingType);
```
Comment (chutten) [inline glean-core/preview/src/metrics/ping.rs]: leftovers?

File: glean-core/preview/src/metrics/ping.rs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+pub struct PingType {
+    pub(crate) name: String,
+    pub(crate) handle: u64,
+}
+
+impl PingType {
+    /// Create a new ping type for the given name and whether to include the client ID when
+    /// sending.
```
Comment (chutten) [inline glean-core/preview/src/metrics/ping.rs]: Doesn't mention `send_if_empty`

**Pull request merged**